app/src/lib/branch.ts

```diff
@@ -1,25 +1,41 @@
 import { Branch, BranchType } from '../models/branch'
 import { UpstreamRemoteName } from './stores'
+import {
+  RepositoryWithGitHubRepository,
+  getNonForkGitHubRepository,
+} from '../models/repository'
 
 /**
- * Finds the remote branch for a branch in the upstream repository
+ * Finds the default branch of the upstream repository of the passed repository.
  *
- * For example:
- * If official/funnel has the branch `development`, then running
- * this on the branches from the fork outofambit/funnel with
- * `development` as the specified branch name will find the
- * branch `remotes/upstream/development`
+ * When the passed repository is not a fork or the fork is configured to use itself
+ * as the ForkContributionTarget, then this method will return null. Otherwise it'll
+ * return the default branch of the upstream repository.
  *
- * @param branchName short name of the branch in the upstream repo
- * @param branches all the branches in the local repo
+ * @param repository The repository to use.
+ * @param branches all the branches in the local repo.
  */
-export function findUpstreamRemoteBranch(
-  branchName: string,
+export function findDefaultUpstreamBranch(
+  repository: RepositoryWithGitHubRepository,
   branches: ReadonlyArray<Branch>
 ) {
+  const githubRepository = getNonForkGitHubRepository(repository)
+
+  // This is a bit hacky... we checked if the result of calling
+  // getNonForkGitHubRepository() is the same as the associated
+  // gitHubRepository. When this happens, we know that either the
+  // repository is not a fork or that it's a fork with
+  // ForkContributionTarget=Self.
+  //
+  // TODO: Make this method return the default branch of the
+  // origin repository instead of null in this scenario.
```

Comment:
This works fine but it's not elegant... I'll fix this by refactoring this method in a separate commit. I started doing that but I need to change quite a few parts of the application since I need to make this method async.

Ideally, we should be able to unify the logic in `infer-comparison-branch.ts` with this one and create a generic function that always returns the default branch of a `Repository` object.